environment/template.go

```diff
@@ -0,0 +1,315 @@
+	FieldName            = "name"
+	FieldStatus          = "status"
+	FieldResourceVersion = "resourceVersion"
+	FieldParameters      = "parameters"
+
+	ValKindTemplate               = "Template"
+	ValKindNamespace              = "Namespace"
+	ValKindConfigMap              = "ConfigMap"
+	ValKindLimitRange             = "LimitRange"
+	ValKindProject                = "Project"
+	ValKindProjectRequest         = "ProjectRequest"
+	ValKindPersistenceVolumeClaim = "PersistentVolumeClaim"
+	ValKindService                = "Service"
+	ValKindSecret                 = "Secret"
+	ValKindServiceAccount         = "ServiceAccount"
+	ValKindRoleBindingRestriction = "RoleBindingRestriction"
+	ValKindRoleBinding            = "RoleBinding"
+	ValKindRoute                  = "Route"
+	ValKindJob                    = "Job"
+	ValKindList                   = "List"
+	ValKindDeployment             = "Deployment"
+	ValKindDeploymentConfig       = "DeploymentConfig"
+	ValKindResourceQuota          = "ResourceQuota"
+
+	varUserName              = "USER_NAME"
+	varProjectUser           = "PROJECT_USER"
+	varProjectRequestingUser = "PROJECT_REQUESTING_USER"
+	varProjectAdminUser      = "PROJECT_ADMIN_USER"
+	varKeycloakURL           = "KEYCLOAK_URL"
+	varCommit                = "COMMIT"
+	varKeycloakOsoEndpoint   = "KEYCLOAK_OSO_ENDPOINT"
+	varKeycloakGHEndpoint    = "KEYCLOAK_GITHUB_ENDPOINT"
+)
+
+var sortOrder = map[string]int{
+	"Namespace":              1,
+	"ProjectRequest":         1,
+	"RoleBindingRestriction": 2,
+	"LimitRange":             3,
+	"ResourceQuota":          4,
+	"Secret":                 5,
+	"ServiceAccount":         6,
+	"Service":                7,
+	"RoleBinding":            8,
+	"PersistentVolumeClaim":  9,
+	"ConfigMap":              10,
+	"DeploymentConfig":       11,
+	"Route":                  12,
+	"Job":                    13,
+}
+
+type Objects []map[interface{}]interface{}
+type Object map[interface{}]interface{}
+
+type Template struct {
+	Filename      string
+	DefaultParams map[string]string
+	Content       string
+	Version       string
+}
+
+var (
+	stageParams       = map[string]string{"DEPLOY_TYPE": "stage"}
+	runParams         = map[string]string{"DEPLOY_TYPE": "run"}
+	noParams          map[string]string
+	specialCharRegexp = regexp.MustCompile("[^a-z0-9]")
+	variableRegexp    = regexp.MustCompile(`\${([A-Z_0-9]+)}`)
+)
+
+func newTemplate(filename string, defaultParams map[string]string) *Template {
+	return &Template{
+		Filename:      filename,
+		DefaultParams: defaultParams,
+	}
+}
+
+func (t *Template) Process(vars map[string]string) (Objects, error) {
+	var objects Objects
+	templateVars := merge(vars, t.DefaultParams)
+	paramsFromTemplate, err := t.getParamsFromTemplate()
+	if err != nil {
+		return objects, err
+	}
+	if paramsFromTemplate != nil {
+		templateVars = merge(paramsFromTemplate, templateVars)
+	}
+	pt, err := t.ReplaceVars(templateVars)
+	if err != nil {
+		return objects, err
+	}
+	t.Version = vars[varCommit]
```

Comment:
As discussed in sync meeting, it would be better if we have latest CommitSHA for given template_file rather latest CommitSHA of tenant_service.  Also, check when you getting CommitSHA corresponding to template_file, you can get commit_timestamp also.  If so then rather using CommitSHA, we can consider using commit_timestamp which will be more human_readable as well comparable (wrt previous and latest versioning).